fsgrid.hpp

```diff
@@ -26,99 +26,181 @@
+      int myRank, MPI_flag;
+      MPI_Initialized(&MPI_flag);
+      if(MPI_flag){
+         MPI_Comm_rank(MPI_COMM_WORLD, &myRank); // TODO allow for separate communicator
+      } else {
+         myRank = FS_MASTER_RANK;
+      }
+
+      std::array<FsSize_t, 3> systemDim;
+      std::array<FsSize_t, 3> processBox;
+      std::array<FsSize_t, 3> minDomainSize;
+      int64_t optimValue = std::numeric_limits<int64_t>::max();
+      std::vector<std::pair<int64_t,std::array<Task_t,3>>> scored_decompositions;
+      scored_decompositions.push_back(std::pair<int64_t,std::array<Task_t,3>>(optimValue,{0,0,0}));
+      for(int i = 0; i < 3; i++) {
+         systemDim[i] = GlobalSize[i];
+         if(GlobalSize[i] == 1) {
+            // In 2D simulation domains, the "thin" dimension can be a single cell thick.
+            minDomainSize[i] = 1;
+         } else {
+            // Otherwise, it needs to be at least as large as our ghost
+            // stencil, so that ghost communication remains consistent.
+            minDomainSize[i] = stencilSize;
          }
-         processDomainDecomposition = {1, 1, 1};
-         for (int i = 1; i <= std::min(nProcs, GlobalSize[0]); i++) {
-            processBox[0] = std::max(systemDim[0]/i, (double)minDomainSize[0]);
-            for (int j = 1; j <= std::min(nProcs, GlobalSize[1]) ; j++) {
-               if( i * j  > nProcs )
-                  break;
-               processBox[1] = std::max(systemDim[1]/j, (double)minDomainSize[1]);
-               for (int k = 1; k <= std::min(nProcs, GlobalSize[2]); k++) {
-                  if( i * j * k > nProcs )
-                     break;
-                  processBox[2] = std::max(systemDim[2]/k, (double)minDomainSize[2]);
-                  double value = 
-                     10 * processBox[0] * processBox[1] * processBox[2] + 
-                     (i > 1 ? processBox[1] * processBox[2]: 0) +
-                     (j > 1 ? processBox[0] * processBox[2]: 0) +
-                     (k > 1 ? processBox[0] * processBox[1]: 0);
-
-                  if(value < optimValue ){
-                     optimValue = value;
-                     processDomainDecomposition[0] = i;
-                     processDomainDecomposition[1] = j;
-                     processDomainDecomposition[2] = k;
-                  }
+      }
+      processDomainDecomposition = {1, 1, 1};
+      for (Task_t i = 1; i <= std::min(nProcs, (Task_t)(GlobalSize[0]/minDomainSize[0])); i++) {
+         for (Task_t j = 1; j <= std::min(nProcs, (Task_t)(GlobalSize[1]/minDomainSize[1])) ; j++) {
+            if( i * j  > nProcs ){
+               break;
+            }
+            Task_t k = nProcs/(i*j);
+            // No need to optimize an incompatible DD, also checks for missing remainders
+            if( i * j * k != nProcs ) {
+               continue;
+            }
+
```

Comment:
So to solve the issue of too small paritioning (what you saw when applying the singleton dimension preference), I think here we should check and continue if `GlobalSize[2]/k < minDomainSize[2]`

Reply:
Oh hey right, getting rid of the last `for` did also get rid of that check.